sale_order_partner_restrict/models/sale.py

```diff
@@ -0,0 +1,57 @@
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    sale_order_partner_restrict = fields.Selection(
+        related="company_id.sale_order_partner_restrict",
+        string="Partner Restriction on Sale Orders",
+    )
+    available_partner_ids = fields.Many2many(
+        "res.partner", compute="_compute_available_partner_ids"
+    )
+
+    @api.depends("partner_id", "sale_order_partner_restrict")
+    def _compute_available_partner_ids(self):
```

Comment:
A fallback is needed in every compute method, otherwise an error is raised if you remove the company from the order.